File: 1612-avoid-flood-in-the-city/1612-avoid-flood-in-the-city.cpp
class Solution {
public:
    vector<int> avoidFlood(vector<int>& rains) {
        int n = rains.size();
        vector<int> ans(n, 1);
        unordered_map<int, int> full;
        set<int> dry_days; 
        for (int i = 0; i < n; i++) {
            if (rains[i] > 0) {
                int lake = rains[i];
                if (full.count(lake)) {
                    auto it = dry_days.upper_bound(full[lake]);
                    if (it == dry_days.end()) {
                        return {}; 
                    }
                    ans[*it] = lake; 
                    dry_days.erase(it);
                }
            
                full[lake] = i;
                ans[i] = -1; 
            } else {
                dry_days.insert(i);
            }
        }
        return ans;
    }
};

File: 1612-avoid-flood-in-the-city/1612-avoid-flood-in-the-city.ts
function avoidFlood(rains: number[]): number[] {
      const n = rains.length;
  const mp = new Map();
  const sunnyDays = [];
  const ans = new Array(n).fill(0);

  for (let day = 0; day < n; day++) {
    const lake = rains[day];

    if (lake > 0) {
      if (mp.has(lake)) {
        const lastRainyDayForLake = mp.get(lake);

        let bestReplacement = -1;
        for (let sunnyDayIndex = 0; sunnyDayIndex < sunnyDays.length; sunnyDayIndex++) {
          if (sunnyDays[sunnyDayIndex] > lastRainyDayForLake) {
            if (bestReplacement === -1 || sunnyDays[sunnyDayIndex] < sunnyDays[bestReplacement]) {
              bestReplacement = sunnyDayIndex;
            }
          }
        }

        if (bestReplacement === -1) {
          return [];
        }

        ans[ sunnyDays[bestReplacement]] = lake;
        sunnyDays.splice(bestReplacement, 1);
      }

      mp.set(lake, day);
      ans[day] = -1;
    } else {
      sunnyDays.push(day);
    }
  }  
  for (const it of sunnyDays) {
    ans[it] = 1;
  }
  return ans;

};